snuba/subscriptions/consumer.py

```diff
@@ -0,0 +1,256 @@
+        # consumed from them will have a value of ``None``.
+        # 2. Partitions that have had a single message consumed will have a
+        # value of ``Position``.
+        # 3. Partitions that have had more than one message consumed will have
+        # a value of ``Interval``.
+        #
+        # Take this example, where a partition contains three messages (MA, MB,
+        # MC) and a scheduled task (T1-TN).
+        #
+        #    Messages:           MA        MB        MC
+        #    Timeline: +---------+---------+---------+---------
+        #    Tasks:    ^    ^    ^    ^    ^    ^    ^    ^
+        #              T1   T2   T3   T4   T5   T6   T7   T8
+        #
+        # In this example, when we are assigned the partition, the state is set
+        # to ``None``. After consuming Message A ("MA"), the partition state
+        # becomes the Position derived from MA. No tasks will have yet been
+        # executed.
+        #
+        # When Message B is consumed, the partition state becomes the
+        # ``Interval`` instance derived from ``(MA, MB)``. At this point, T4
+        # and T5 (the tasks that are scheduled between the timestamps of
+        # messages "MA" and "MB") will be included in the ``TaskSet`` returned
+        # by the ``poll`` call. T3 will not be included, since it was
+        # presumably contained within a ``TaskSet`` instance returned by a
+        # previous ``poll`` call. The lower bound ("MA" in this case) is
+        # exclusive, while the upper bound ("MB") is inclusive.
+        #
+        # When all tasks in the ``TaskSet`` have been successfully evaluated,
+        # committing the task set will commit the *lower bound* offset of this
+        # task set. The lower bound is selected so that on consumer restart or
+        # rebalance, the message that has an offset greater than the lower
+        # bound (in our case, "MB") will be the first message consumed. The
+        # next tasks to be executed will be those that are scheduled between
+        # the timestamps of "MB" and "MC" (again: lower bound exclusive, upper
+        # bound inclusive): T6 and T7.
+        self.__topic_partition_states: MutableMapping[
+            Tuple[str, int], PartitionState
+        ] = {}
+
+        # Using the built-in ``enable.auto.offset.store`` behavior will cause
+        # gaps in task execution during restarts or rebalances, since it will
+        # set the next offset to be consumed as the most recently message
+        # consumed's offset + 1. In our case, we always want the first message
+        # to be consumed after a rebalance to be *the same* as the previously
+        # consumed message. To account for this, we have to always disable it,
+        # and use our own implementation that stores the offset of the most
+        # recently received message in a partition, instead of that offset + 1.
+        self.__consumer = KafkaConsumer(
```

Comment:
This will eventually need to act similarly to the `SynchronizedConsumer`, since we shouldn't be executing queries that haven't been written yet. That does mean we'll have to extract the `SynchronizedConsumer` from Sentry at some point (either by copying, or putting it in a library.)